File: frontend/lib/achievements.ts
import type { Place } from "@/lib/places";

export const JUAN_DE_FUCA_PARK_ID = "provincial-juan-de-fuca-park";

type AchievementContext = {
  places: Place[];
  visited: ReadonlySet<string>;
  visitTimestamps?: Readonly<Record<string, string>>;
};

export type Achievement = {
  id: string;
  name: string;
  species: string;
  description: string;
  requiredPlaceIds?: readonly string[];
  current: number;
  target: number;
  earned: boolean;
  earnedAt?: string;
};

type Definition = Omit<Achievement, "current" | "target" | "earned"> & {
  target: number;
  count: (context: AchievementContext) => number;
};

const total = ({ places, visited }: AchievementContext) => places.filter((place) => visited.has(place.id)).length;
const regionCount = (needle: string) => ({ places, visited }: AchievementContext) =>
  places.filter((place) => visited.has(place.id) && place.region.toLocaleLowerCase().includes(needle)).length;
const placeSetCount = (ids: readonly string[]) => ({ places, visited }: AchievementContext) => {
  const active = new Set(places.map((place) => place.id));
  return ids.filter((id) => active.has(id) && visited.has(id)).length;
};

const bananaSlugPlaces = [JUAN_DE_FUCA_PARK_ID, "provincial-carmanah-walbran-park", "provincial-macmillan-park"] as const;
const blackBearCoastPlaces = ["national-pacific-rim-national-park-reserve", "provincial-cape-scott-park", "provincial-strathcona-park"] as const;
const seaOtterRaftPlaces = ["island-nootka-island", "island-flores-island", "island-vargas-island"] as const;
const orcaLookoutPlaces = ["national-gulf-islands-national-park-reserve", "regional-east-point-regional-park", "regional-brooks-point-regional-park"] as const;
const stellerHighCountryPlaces = ["provincial-strathcona-park", "provincial-schoen-lake-park", "regional-mount-arrowsmith-massif-regional-park"] as const;
const deerSouthIslandPlaces = ["regional-east-sooke-regional-park", "provincial-gowlland-tod-park", "provincial-goldstream-park"] as const;
const treefrogPondHopPlaces = ["regional-elk-beaver-lake-regional-park", "regional-matheson-lake-regional-park", "regional-thetis-lake-regional-park"] as const;
const redLeggedWetlandsPlaces = ["provincial-kennedy-river-bog-park", "regional-coats-marsh-regional-park", "regional-hamilton-marsh-regional-park-and-conservation-area"] as const;
const redcedarRainlinePlaces = ["national-pacific-rim-national-park-reserve", "provincial-carmanah-walbran-park", JUAN_DE_FUCA_PARK_ID] as const;
const arbutusRainshadowPlaces = ["provincial-bodega-ridge-park", "provincial-helliwell-park", "provincial-mount-maxwell-park"] as const;
const swordFernFallsPlaces = ["provincial-elk-falls-park", "provincial-englishman-river-falls-park", "provincial-little-qualicum-falls-park"] as const;
const camasRainshadowPlaces = ["national-gulf-islands-national-park-reserve", "provincial-ruckle-park", "provincial-mount-maxwell-park"] as const;
const heronTidelinePlaces = ["regional-island-view-beach-regional-park", "regional-witty-s-lagoon-regional-park", "regional-little-qualicum-river-estuary-regional-conservation-area"] as const;
const harbourSealShorelinePlaces = ["provincial-rathtrevor-beach-park", "provincial-miracle-beach-park", "provincial-french-beach-park"] as const;

const definitions: Definition[] = [
  { id: "banana-slug-medal", name: "Banana Slug Rainwalk", species: "banana-slug", description: "Visit Juan de Fuca, Carmanah Walbran and Macmillan parks.", requiredPlaceIds: bananaSlugPlaces, target: 3, count: placeSetCount(bananaSlugPlaces) },
  { id: "black-bear-coast", name: "Black Bear Coast", species: "black-bear", description: "Visit Pacific Rim, Cape Scott and Strathcona parks.", requiredPlaceIds: blackBearCoastPlaces, target: 3, count: placeSetCount(blackBearCoastPlaces) },
  { id: "sea-otter-raft", name: "Sea Otter Raft", species: "sea-otter", description: "Visit Nootka, Flores and Vargas islands along the outer coast.", requiredPlaceIds: seaOtterRaftPlaces, target: 3, count: placeSetCount(seaOtterRaftPlaces) },
  { id: "orca-salish-lookouts", name: "Orca Lookout Loop", species: "orca", description: "Visit Gulf Islands National Park Reserve, East Point and Brooks Point.", requiredPlaceIds: orcaLookoutPlaces, target: 3, count: placeSetCount(orcaLookoutPlaces) },
  { id: "river-otter-rookie", name: "River Otter Rookie", species: "river-otter", description: "Collect your first place.", target: 1, count: total },
  { id: "harbour-seal-five", name: "Harbour Seal High Five", species: "harbour-seal", description: "Visit 5 places.", target: 5, count: total },
  { id: "eagle-ten", name: "Bald Eagle Ten", species: "bald-eagle", description: "Visit 10 places.", target: 10, count: total },
  { id: "heron-twenty-five", name: "Heron’s Long Stride", species: "great-blue-heron", description: "Visit 25 places.", target: 25, count: total },
  { id: "kingfisher-fifty", name: "Kingfisher Fifty", species: "belted-kingfisher", description: "Visit 50 places.", target: 50, count: total },
  { id: "hummingbird-century", name: "Hummingbird Century", species: "rufous-hummingbird", description: "Visit 100 places.", target: 100, count: total },
  { id: "steller-high-country", name: "Steller’s Jay High Country", species: "stellers-jay", description: "Visit Strathcona, Schoen Lake and Mount Arrowsmith Massif parks.", requiredPlaceIds: stellerHighCountryPlaces, target: 3, count: placeSetCount(stellerHighCountryPlaces) },
  { id: "deer-south-island", name: "Black-tailed Deer Ramble", species: "black-tailed-deer", description: "Visit East Sooke, Gowlland Tod and Goldstream parks.", requiredPlaceIds: deerSouthIslandPlaces, target: 3, count: placeSetCount(deerSouthIslandPlaces) },
  { id: "treefrog-pond-hop", name: "Treefrog Pond Hop", species: "pacific-treefrog", description: "Visit Elk/Beaver Lake, Matheson Lake and Thetis Lake regional parks.", requiredPlaceIds: treefrogPondHopPlaces, target: 3, count: placeSetCount(treefrogPondHopPlaces) },
  { id: "red-legged-wetlands", name: "Red-legged Wetland Hop", species: "red-legged-frog", description: "Visit Kennedy River Bog, Coats Marsh and Hamilton Marsh.", requiredPlaceIds: redLeggedWetlandsPlaces, target: 3, count: placeSetCount(redLeggedWetlandsPlaces) },
  { id: "douglas-fir-central", name: "Douglas-fir Heartwood", species: "douglas-fir", description: "Visit 8 places in Central Island.", target: 8, count: regionCount("central") },
  { id: "redcedar-rainline", name: "Western Redcedar Rainline", species: "western-redcedar", description: "Visit Pacific Rim, Carmanah Walbran and Juan de Fuca parks.", requiredPlaceIds: redcedarRainlinePlaces, target: 3, count: placeSetCount(redcedarRainlinePlaces) },
  { id: "arbutus-rainshadow", name: "Arbutus Rainshadow", species: "arbutus", description: "Visit Bodega Ridge, Helliwell and Mount Maxwell parks.", requiredPlaceIds: arbutusRainshadowPlaces, target: 3, count: placeSetCount(arbutusRainshadowPlaces) },
  { id: "salal-north", name: "Salal Northbound", species: "salal", description: "Visit 5 places in a northern region.", target: 5, count: regionCount("northern") },
  { id: "sword-fern-falls", name: "Sword Fern Falls", species: "sword-fern", description: "Visit Elk Falls, Englishman River Falls and Little Qualicum Falls.", requiredPlaceIds: swordFernFallsPlaces, target: 3, count: placeSetCount(swordFernFallsPlaces) },
  { id: "camas-rainshadow", name: "Camas Rainshadow", species: "camas", description: "Visit Gulf Islands National Park Reserve, Ruckle and Mount Maxwell parks.", requiredPlaceIds: camasRainshadowPlaces, target: 3, count: placeSetCount(camasRainshadowPlaces) },
  { id: "heron-tideline", name: "Heron Tideline", species: "great-blue-heron", description: "Visit Island View Beach, Witty’s Lagoon and Little Qualicum River Estuary.", requiredPlaceIds: heronTidelinePlaces, target: 3, count: placeSetCount(heronTidelinePlaces) },
  { id: "harbour-seal-shores", name: "Harbour Seal Shoreline", species: "harbour-seal", description: "Visit Rathtrevor Beach, Miracle Beach and French Beach parks.", requiredPlaceIds: harbourSealShorelinePlaces, target: 3, count: placeSetCount(harbourSealShorelinePlaces) },
];

export function achievements(context: AchievementContext): Achievement[] {
  return definitions.map(({ count, target, ...definition }) => {
    const current = Math.min(count(context), target);
    const earned = current >= target;
    const chronology = [...context.visited]
      .map((id) => ({ id, timestamp: context.visitTimestamps?.[id] }))
      .filter((entry): entry is { id: string; timestamp: string } => Boolean(entry.timestamp))
      .sort((a, b) => a.timestamp.localeCompare(b.timestamp));
    const seen = new Set<string>();
    let earnedAt: string | undefined;
    for (const entry of chronology) {
      seen.add(entry.id);
      if (count({ ...context, visited: seen }) >= target) {
        earnedAt = entry.timestamp;
        break;
      }
    }
    return { ...definition, current, target, earned, ...(earnedAt ? { earnedAt } : {}) };
  });
}

export function newlyEarnedAchievementIds(previous: readonly Achievement[], next: readonly Achievement[]): string[] {
  const previouslyEarned = new Set(previous.filter((badge) => badge.earned).map((badge) => badge.id));
  return next.filter((badge) => badge.earned && !previouslyEarned.has(badge.id)).map((badge) => badge.id);
}

File: backend/app/achievements.py
"""Server-side achievement rules mirrored from the web application's catalog."""

from __future__ import annotations

from datetime import datetime


_DEFINITIONS = (
    ("banana-slug-medal", "Banana Slug Rainwalk", "banana-slug", "Visit Juan de Fuca, Carmanah Walbran and Macmillan parks.", ("provincial-juan-de-fuca-park", "provincial-carmanah-walbran-park", "provincial-macmillan-park"), 3, "places"),
    ("black-bear-coast", "Black Bear Coast", "black-bear", "Visit Pacific Rim, Cape Scott and Strathcona parks.", ("national-pacific-rim-national-park-reserve", "provincial-cape-scott-park", "provincial-strathcona-park"), 3, "places"),
    ("sea-otter-raft", "Sea Otter Raft", "sea-otter", "Visit Nootka, Flores and Vargas islands along the outer coast.", ("island-nootka-island", "island-flores-island", "island-vargas-island"), 3, "places"),
    ("orca-salish-lookouts", "Orca Lookout Loop", "orca", "Visit Gulf Islands National Park Reserve, East Point and Brooks Point.", ("national-gulf-islands-national-park-reserve", "regional-east-point-regional-park", "regional-brooks-point-regional-park"), 3, "places"),
    ("river-otter-rookie", "River Otter Rookie", "river-otter", "Collect your first place.", (), 1, "places"),
    ("harbour-seal-five", "Harbour Seal High Five", "harbour-seal", "Visit 5 places.", (), 5, "places"),
    ("eagle-ten", "Bald Eagle Ten", "bald-eagle", "Visit 10 places.", (), 10, "places"),
    ("heron-twenty-five", "Heron’s Long Stride", "great-blue-heron", "Visit 25 places.", (), 25, "places"),
    ("kingfisher-fifty", "Kingfisher Fifty", "belted-kingfisher", "Visit 50 places.", (), 50, "places"),
    ("hummingbird-century", "Hummingbird Century", "rufous-hummingbird", "Visit 100 places.", (), 100, "places"),
    ("steller-high-country", "Steller’s Jay High Country", "stellers-jay", "Visit Strathcona, Schoen Lake and Mount Arrowsmith Massif parks.", ("provincial-strathcona-park", "provincial-schoen-lake-park", "regional-mount-arrowsmith-massif-regional-park"), 3, "places"),
    ("deer-south-island", "Black-tailed Deer Ramble", "black-tailed-deer", "Visit East Sooke, Gowlland Tod and Goldstream parks.", ("regional-east-sooke-regional-park", "provincial-gowlland-tod-park", "provincial-goldstream-park"), 3, "places"),
    ("treefrog-pond-hop", "Treefrog Pond Hop", "pacific-treefrog", "Visit Elk/Beaver Lake, Matheson Lake and Thetis Lake regional parks.", ("regional-elk-beaver-lake-regional-park", "regional-matheson-lake-regional-park", "regional-thetis-lake-regional-park"), 3, "places"),
    ("red-legged-wetlands", "Red-legged Wetland Hop", "red-legged-frog", "Visit Kennedy River Bog, Coats Marsh and Hamilton Marsh.", ("provincial-kennedy-river-bog-park", "regional-coats-marsh-regional-park", "regional-hamilton-marsh-regional-park-and-conservation-area"), 3, "places"),
    ("douglas-fir-central", "Douglas-fir Heartwood", "douglas-fir", "Visit 8 places in Central Island.", (), 8, "region:central"),
    ("redcedar-rainline", "Western Redcedar Rainline", "western-redcedar", "Visit Pacific Rim, Carmanah Walbran and Juan de Fuca parks.", ("national-pacific-rim-national-park-reserve", "provincial-carmanah-walbran-park", "provincial-juan-de-fuca-park"), 3, "places"),
    ("arbutus-rainshadow", "Arbutus Rainshadow", "arbutus", "Visit Bodega Ridge, Helliwell and Mount Maxwell parks.", ("provincial-bodega-ridge-park", "provincial-helliwell-park", "provincial-mount-maxwell-park"), 3, "places"),
    ("salal-north", "Salal Northbound", "salal", "Visit 5 places in a northern region.", (), 5, "region:northern"),
    ("sword-fern-falls", "Sword Fern Falls", "sword-fern", "Visit Elk Falls, Englishman River Falls and Little Qualicum Falls.", ("provincial-elk-falls-park", "provincial-englishman-river-falls-park", "provincial-little-qualicum-falls-park"), 3, "places"),
    ("camas-rainshadow", "Camas Rainshadow", "camas", "Visit Gulf Islands National Park Reserve, Ruckle and Mount Maxwell parks.", ("national-gulf-islands-national-park-reserve", "provincial-ruckle-park", "provincial-mount-maxwell-park"), 3, "places"),
    ("heron-tideline", "Heron Tideline", "great-blue-heron", "Visit Island View Beach, Witty’s Lagoon and Little Qualicum River Estuary.", ("regional-island-view-beach-regional-park", "regional-witty-s-lagoon-regional-park", "regional-little-qualicum-river-estuary"), 3, "places"),
    ("harbour-seal-shores", "Harbour Seal Shoreline", "harbour-seal", "Visit Rathtrevor Beach, Miracle Beach and French Beach parks.", ("provincial-rathtrevor-beach-park", "provincial-miracle-beach-park", "provincial-french-beach-park"), 3, "places"),
)


def _count(
    definition: tuple,
    *,
    active_ids: set[str],
    visited_ids: set[str],
    regions: dict[str, str],
) -> int:
    _, _, _, _, required_ids, _, kind = definition
    if kind == "places":
        if required_ids:
            return sum(
                1 for place_id in required_ids
                if place_id in active_ids and place_id in visited_ids
            )
        return len(active_ids & visited_ids)
    region_name = kind.removeprefix("region:")
    return sum(
        1 for place_id in active_ids & visited_ids
        if region_name in regions.get(place_id, "").casefold()
    )


def achievements(
    *,
    places: list[dict],
    visits: list[dict],
) -> list[dict]:
    """Calculate badge progress and earning timestamp from visible catalogue rows."""

    active_ids = {place["id"] for place in places}
    regions = {place["id"]: place["region"] for place in places}
    names = {place["id"]: place["name"] for place in places}
    visit_by_id = {
        visit["place_id"]: visit["visited_at"]
        for visit in visits
        if visit["place_id"] in active_ids
    }
    visited_ids = set(visit_by_id)
    chronology = sorted(
        visit_by_id.items(),
        key=lambda item: (item[1], item[0]),
    )
    results: list[dict] = []
    for definition in _DEFINITIONS:
        badge_id, name, species, description, required_ids, target, _ = definition
        current = min(
            _count(
                definition,
                active_ids=active_ids,
                visited_ids=visited_ids,
                regions=regions,
            ),
            target,
        )
        earned_at: datetime | None = None
        if current >= target:
            seen: set[str] = set()
            for place_id, timestamp in chronology:
                seen.add(place_id)
                if _count(
                    definition,
                    active_ids=active_ids,
                    visited_ids=seen,
                    regions=regions,
                ) >= target:
                    earned_at = timestamp
                    break
        result = {
            "id": badge_id,
            "name": name,
            "species": species,
            "description": description,
            "current": current,
            "target": target,
            "earned": current >= target,
        }
        if required_ids:
            result["required_place_ids"] = list(required_ids)
            result["required_places"] = [
                {"id": place_id, "name": names[place_id]}
                for place_id in required_ids
                if place_id in names
            ]
        if earned_at is not None:
            result["earned_at"] = earned_at
        results.append(result)
    return results
